lib/ansible/modules/crypto/openssh_keypair.py

```diff
@@ -0,0 +1,351 @@
+            if self.size not in (256, 384, 521):
+                module.fail_json(msg='''For ECDSA keys, size determines the key length by selecting from
+                                     one of three elliptic curve sizes: 256, 384 or 521 bits.
+                                     Attempting to use bit lengths other than these three values for
+                                     ECDSA keys will cause this module to fail.''')
+        if self.type == 'ed25519':
+            self.size = 256
+
+    def generate(self, module):
+        # generate a keypair
+        if not self.check(module, perms_required=False) or self.force:
+            args = [
+                'ssh-keygen',
+                '-q',
+                '-b', str(self.size),
+                '-t', self.type,
+                '-f', self.path,
+            ]
+            try:
+                tmpfile = None
+                if self.passphrase:
+                    with tempfile.NamedTemporaryFile(delete=False) as tf:
+                        tmpfile = tf.name
+                        module.add_cleanup_file(tmpfile)
+                        if sys.version_info[0] < 3:
+                            tf.write("#!/bin/sh\necho %s" % quote(self.passphrase))
+                    if sys.version_info[0] >= 3:
+                        with io.open(tmpfile, "w", encoding='utf-8') as ft:
+                            ft.write("#!/bin/sh\necho %s" % quote(self.passphrase))
+                    env = {"DISPLAY": "fake", "SSH_ASKPASS": str(tmpfile)}
+                else:
+                    args.extend(['-N', ''])
+                    env = {}
+
+                if self.comment:
+                    args.extend(['-C', self.comment])
+                else:
+                    args.extend(['-C', ""])
+
+                self.changed = True
+                subprocess.call(args, env=env)
+
+                proc = Popen(['ssh-keygen', '-lf', self.path], stdout=PIPE)
+                self.fingerprint = proc.communicate()[0].split()
+            except Exception as e:
+                self.remove()
+                module.fail_json(msg="%s" % to_native(e))
+
+        file_args = module.load_file_common_arguments(module.params)
+        if module.set_fs_attributes_if_different(file_args, False):
+            self.changed = True
+
+    def check(self, module, perms_required=True):
+
+        # check if the key is correct
+        def _check_state():
+            return os.path.exists(self.path)
+
+        if _check_state():
+            proc = Popen(['ssh-keygen', '-lf', self.path], stdout=PIPE)
+            fingerprint = proc.communicate()[0].split()
+            keysize = int(fingerprint[0])
+            keytype = fingerprint[-1][1:-1].lower()
+        else:
+            return False
+
+        def _check_perms(module):
+            file_args = module.load_file_common_arguments(module.params)
+            return not module.set_fs_attributes_if_different(file_args, False)
```

Comment:
This gets called in check mode too and might change attributes on disk

Reply:
Looks like an issue in https://github.com/ansible/ansible/blob/8a6059ad0a5c9e96adb24f9dec1da9687a415045/lib/ansible/module_utils/crypto.py#L129-L131 too

Reply:
Ah thanks, I checked the implementation again, it verifies this here:

https://github.com/ansible/ansible/blob/f2dccb90e893df30c2b8bfc925dba80f6ceed6a7/lib/ansible/module_utils/basic.py#L1376